webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/internal/NodeDetailsAPI.scala

```diff
@@ -0,0 +1,149 @@
+  , restExtractor : RestExtractorService
+  , reportsExecutionRepository: RoReportsExecutionRepository
+  , getGlobalMode : () => Box[GlobalPolicyMode]
+  , readOnlySoftwareDAO: ReadOnlySoftwareDAO
+
+) extends  RestHelper with  Loggable {
+
+
+  def serialize(agentRunWithNodeConfig: Option[AgentRunWithNodeConfig], globalPolicyMode: GlobalPolicyMode, nodeInfo : NodeInfo, properties : List[String], softs: List[Software]) = {
+    import net.liftweb.json.JsonDSL._
+
+    val (policyMode,explanation) =
+      (globalPolicyMode.overridable,nodeInfo.policyMode) match {
+        case (Always,Some(mode)) =>
+          (mode,"<p>This mode is an override applied to this node. You can change it in the <i><b>node's settings</b></i>.</p>")
+        case (Always,None) =>
+          val expl = """<p>This mode is the globally defined default. You can change it in <i><b>settings</b></i>.</p><p>You can also override it on this node in the <i><b>node's settings</b></i>.</p>"""
+          (globalPolicyMode.mode, expl)
+        case (Unoverridable,_) =>
+          (globalPolicyMode.mode, "<p>This mode is the globally defined default. You can change it in <i><b>Settings</b></i>.</p>")
+      }
+    (  ("name" -> nodeInfo.hostname)
+      ~  ("policyServerId" -> nodeInfo.policyServerId.value)
+      ~  ("policyMode" -> policyMode.name)
+      ~  ("explanation" -> explanation)
+      ~  ("kernel" -> nodeInfo.osDetails.kernelVersion.value)
+      ~  ("agentVersion" -> nodeInfo.agentsName.headOption.flatMap(_.version.map(_.value)))
+      ~  ("id" -> nodeInfo.id.value)
+      ~  ("ram" -> nodeInfo.ram.map(_.toStringMo()))
+      ~  ("machineType" -> nodeInfo.machine.map(_.machineType.toString))
+      ~  ("os" -> nodeInfo.osDetails.fullName)
+      ~  ("state" -> nodeInfo.state.name)
+      ~  ("ipAddresses" -> nodeInfo.ips.filter(ip => ip != "127.0.0.1" && ip != "0:0:0:0:0:0:0:1"))
+      ~  ("lastRun" -> agentRunWithNodeConfig.map(d => DateFormaterService.getDisplayDate(d.agentRunId.date)).getOrElse("Never"))
+      ~  ("software" -> JObject(softs.toList.map(s => JField(s.name.getOrElse(""), JString(s.version.map(_.value).getOrElse("N/A"))))))
+      ~  ("property" -> JObject(nodeInfo.properties.filter(p => properties.contains(p.name)).map(p => JField(p.name, parse(p.value.render(ConfigRenderOptions.concise()) ) )) ))
+      )
+  }
+  def requestDispatch: PartialFunction[Req, () => Box[LiftResponse]] = {
+    case Post("software" :: software :: Nil , req) =>
+      import com.normation.box._
+
+      val n1 = System.currentTimeMillis
+      for {
+
+        optNodeIds <- req.json.flatMap(restExtractor.extractNodeIdsFromJson)
+
+        nodes <- optNodeIds match {
+          case None => nodeInfoService.getAll()
+          case Some(nodeIds) => com.normation.utils.Control.sequence(nodeIds)( nodeInfoService.getNodeInfo(_).map(_.map(n => (n.id, n)))).map(_.flatten.toMap)
+        }
+        softs <- readOnlySoftwareDAO.getNodesbySofwareName(software).toBox.map(_.toMap)
```

Comment:
I'm not sure this would do what you want, as soft is a list of [NodeId, Software], but there no constraint that there is only one software per node, meaning you can have (Node1, (SoftA,V1)), (Node1, (SoftA,V2))  
(same node, different software version, or publisher, for a same software name)